src/main/java/com/aws/greengrass/clientdevices/auth/util/WildcardTrie.java

```diff
@@ -0,0 +1,220 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.clientdevices.auth.util;
+
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+@RequiredArgsConstructor
+public class WildcardTrie {
+
+    private static final Function<WildcardType, UnsupportedOperationException> EXCEPTION_UNSUPPORTED_WILDCARD_TYPE = wildcardType ->
+            new UnsupportedOperationException("wildcard type " + wildcardType.name() + " not supported");
+
+    private Node root;
+
+    private final MatchOptions opts;
+
+    private static String cleanPattern(@NonNull String s) {
+        // for example "abc***def" can be reduced to "abc*def"
+        return s.replaceAll(String.format("\\%s+", WildcardType.GLOB.val), WildcardType.GLOB.val);
+    }
+
+    public WildcardTrie withPattern(@NonNull String s) {
+        root = new Node();
+        withPattern(root, cleanPattern(s));
+        return this;
+    }
+
+    private Node withPattern(@NonNull Node n, @NonNull String s) {
+        StringBuilder token = new StringBuilder(s.length());
+        for (int i = 0; i < s.length(); i++) {
+            char c = s.charAt(i);
+            if (isWildcard(s.charAt(i))) {
+                // create child node from non-wildcard chars that have been accumulated so far
+                Node node = token.length() > 0 ? n.children.get(token.toString()) : n;
+                // create child node for wildcard char itself
+                WildcardType type = WildcardType.from(c);
+                node = node.children.get(type.val());
+                node.wildcardType = type;
+                if (i == s.length() - 1) {
+                    // we've reached the last token
+                    return node;
+                }
+                return withPattern(node, s.substring(i + 1));
+            } else {
+                token.append(c);
+            }
+        }
+        // use remaining (non-wildcard) chars as last token
+        if (token.length() > 0) {
+            return n.children.get(token.toString());
+        } else {
+            return n;
+        }
+    }
+
+    private boolean isWildcard(char c) {
+        WildcardType type = WildcardType.from(c);
+        if (type == null) {
+            return false;
+        }
+        if (type == WildcardType.SINGLE) {
+            return opts.useSingleCharWildcard;
+        }
+        return true;
+    }
+
+    public boolean matches(@NonNull String s) {
+        if (root == null) {
+            return s.isEmpty();
+        }
+        return matches(root, s);
+    }
+
+    private boolean matches(@NonNull Node n, @NonNull String s) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.